### PR TITLE
Fourth module (Authentication & Spring Security)

### PR DESCRIPTION
**Pull request by PavelSidorovich**

### CODE REVIEW

File: persistence/src/main/java/com/epam/esm/gcs/model/AccountModel_.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package com.epam.esm.gcs.model;
```
Comment (vlad-gal) [inline persistence/src/main/java/com/epam/esm/gcs/model/AccountModel_.java:1]: As I correctly understand, it's autogenerated class, why is it in src package?

File: service/src/main/java/com/epam/esm/gcs/service/OrderService.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+public interface OrderService {
+
+    OrderDto purchase(long userId, long certificateId);
+
+    List<OrderDto> findByUserId(long userId, Pageable pageable);
+
+    TruncatedOrderDto findTruncatedByIds(long userId, long orderId);
+
+//    TagDto findMostWidelyTag();
```
Comment (vlad-gal) [inline service/src/main/java/com/epam/esm/gcs/service/OrderService.java:17]: Please remove or uncomment

File: service/src/main/java/com/epam/esm/gcs/service/UserService.java [hunk 3]
```diff
@@ -1,11 +1,22 @@
...
+
+    UserDto findById(long id);
+
+    List<UserDto> findAll(Pageable pageable);
+
+    UserDto signUp(SignUpUserDto signUpDto);
 
     boolean existsWithEmail(String email);
 
+//    UserDto findTheMostActiveUser();
```
Comment (vlad-gal) [inline service/src/main/java/com/epam/esm/gcs/service/UserService.java:20]: Please remove or uncomment

**Review by vlad-gal — APPROVED** [review]

File: persistence/src/main/java/com/epam/esm/gcs/model/AccountRoleModel_.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
+package com.epam.esm.gcs.model;
+
+import javax.annotation.processing.Generated;
+import javax.persistence.metamodel.SingularAttribute;
+import javax.persistence.metamodel.StaticMetamodel;
+
+@Generated(value = "org.hibernate.jpamodelgen.JPAMetaModelEntityProcessor")
+@StaticMetamodel(AccountRoleModel.class)
+public abstract class AccountRoleModel_ {
```
Comment (ubarouski) [inline persistence/src/main/java/com/epam/esm/gcs/model/AccountRoleModel_.java:9]: Lets discuss on demo

File: persistence/src/main/java/com/epam/esm/gcs/model/CertificateModel.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    @CreatedDate
+    @Column(name = CertificateModel_.CREATE_DATE, nullable = false)
+    private LocalDateTime createDate;
+
+    @LastModifiedDate
+    @Column(name = CertificateModel_.LAST_UPDATE_DATE, nullable = false)
+    private LocalDateTime lastUpdateDate;
+
+    @Exclude
+    @ManyToMany(fetch = FetchType.LAZY)
```
Comment (ubarouski) [inline persistence/src/main/java/com/epam/esm/gcs/model/CertificateModel.java:50]: Lazy is default, isn't it?

File: repository/src/main/java/com/epam/esm/gcs/repository/CertificateRepository.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+public interface CertificateRepository
+        extends JpaRepository<CertificateModel, Long>,
+                JpaSpecificationExecutor<CertificateModel> {
+
+    @Query("SELECT DISTINCT c FROM CertificateModel c " +
+           "JOIN c.tags as t " +
+           "WHERE lower(t.name) IN ?1 " +
+           "group by c.id, c.name, c.price, c.duration, " +
+           "c.description, c.createDate, c.lastUpdateDate " +
+           "having count (t.id) = ?2")
```
Comment (ubarouski) [inline repository/src/main/java/com/epam/esm/gcs/repository/CertificateRepository.java:23]: Why do you need count?

File: repository/src/main/java/com/epam/esm/gcs/repository/TagRepository.java [hunk 7]
```diff
@@ -1,13 +1,33 @@
...
-    boolean existsWithName(String name);
+    boolean existsByNameIgnoreCase(String name);
+
+    @Query(value = "SELECT tag_id as id, ct.name as name\n" +
+                   "FROM user_order\n" +
+                   "         inner join certificate c on c.id = user_order.certificate_id\n" +
+                   "         inner join certificates_by_tags cbt on c.id = cbt.certificate_id\n" +
+                   "         inner join certificate_tag ct on cbt.tag_id = ct.id\n" +
+                   "where user_id = (SELECT user_id\n" +
+                   "                 FROM (SELECT user_id\n" +
```
Comment (ubarouski) [inline repository/src/main/java/com/epam/esm/gcs/repository/TagRepository.java:23]: Is it really necessary to have subquery?

File: service/src/main/java/com/epam/esm/gcs/service/TagService.java [hunk 8]
```diff
@@ -8,4 +8,6 @@
 
     boolean existsWithName(String name);
 
+    TagDto findTheMostUsedTag();
```
Comment (ubarouski) [inline service/src/main/java/com/epam/esm/gcs/service/TagService.java:11]: I think it is better without article in the name findMostUsedTag

File: web/src/main/java/com/epam/esm/gcs/util/impl/JwtTokenUtilImpl.java [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.stereotype.Service;
+
+import java.util.Date;
+import java.util.HashMap;
+import java.util.function.Function;
+
+@Service
+public class JwtTokenUtilImpl implements JwtTokenUtil {
```
Comment (ubarouski) [inline web/src/main/java/com/epam/esm/gcs/util/impl/JwtTokenUtilImpl.java:16]: Is it utility class? If yes why it is not static?

**Review by ubarouski — APPROVED** [review]

**Review by AMiadzvedzeu — APPROVED** [review]

**Pull request merged**